pkg/ctrlz/topics/commands.go

```diff
@@ -0,0 +1,53 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package topics
+
+import (
+	"fmt"
+	"html/template"
+	"net/http"
+
+	"istio.io/istio/pkg/ctrlz/fw"
+	"istio.io/istio/pkg/reloadwatcher"
+)
+
+type commandsTopic struct {
+}
+
+// CommandsTopic returns a ControlZ topic that tools to command signals to the process
+func CommandsTopic() fw.Topic {
+	return commandsTopic{}
+}
+
+func (commandsTopic) Title() string {
+	return "Command"
+}
+
+func (commandsTopic) Prefix() string {
+	return "command"
+}
+
+func (commandsTopic) Activate(context fw.TopicContext) {
+	tmpl := template.Must(context.Layout().Parse(string(MustAsset("assets/templates/commands.html"))))
+
+	_ = context.HTMLRouter().StrictSlash(true).NewRoute().Path("/").HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		fw.RenderHTML(w, tmpl, nil)
+	})
+
+	_ = context.JSONRouter().StrictSlash(true).NewRoute().Methods("PUT", "POST").Path("/reloadconfig").HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		reloadwatcher.Notify(fmt.Sprintf("Remote: %v", req.RemoteAddr))
```

Comment:
So ControlZ now has a dependency on the reloadWatcher?

Reply:
The reload trigger isn't really specific to Galley. Its hooked up there because this is pressing for testing but its needed for Pilot too.

Its essentially an abstraction over os signals handling at this point

Reply:
I think it's more than an abstraction over signals, it's a concrete hard-wired piece of logic that assumes particular behavior, with /reloadconfig and such.

I know we only have a few components in Istio at this point, so arguing about making generic abstractions may seem silly, but it's the house of cards kind of argument here I believe.

I would rather this be in the specific components where component-specific logic is contained. It would make CtrlZ, which is arguably a low-level abstraction, not have funny dependencies and hard-wired behavior. CtrlZ was designed specifically to make it easy for components to add their own topics, as Oz showed in his comments, so let's leverage that.

Reply:
Just as a point of reference at Google we treat handlers like this as common utility in order to encourage conventions around their definition and more importantly their use.

The utility here is pretty clearly generic and I would disagree about 'assuming a particular behavior'. The only thing that implies behavior in the implementation is the naming convention '/reloadconfig' which I'm happy to change though having a convention is exactly the behavior we want to encourage as it leads to a consistent operator experience. Alternative naming suggestions welcome.

Reply:
So if this is an abstraction over OS signals, how about renaming /reloadconfig to a signal-inspired name? Then I think I'd feel better about the whole thing.

Reply:
Done. PTAL